linkerd/opencensus/src/lib.rs

```diff
@@ -1,226 +1,134 @@
 use tracing::trace;
-
 pub mod metrics;
 
+pub async fn export_spans<T, S>(client: T, node: Node, spans: S, metrics: Registry)
+where
+    T: GrpcService<BoxBody> + Clone,
+    T::Error: Into<Error>,
+    <T::ResponseBody as HttpBody>::Error: Into<Error> + Send + Sync,
+    T::ResponseBody: 'static,
+    S: Stream<Item = Span> + Unpin,
+{
+    SpanExporter::new(client, node, spans, metrics).run().await
+}
+
 /// SpanExporter sends a Stream of spans to the given TraceService gRPC service.
-#[pin_project]
-pub struct SpanExporter<T, S> {
+struct SpanExporter<T, S> {
     client: T,
     node: Node,
-    state: State,
-    #[pin]
     spans: S,
-    max_batch_size: usize,
     metrics: Registry,
 }
 
-#[allow(clippy::large_enum_variant)]
-enum State {
-    Idle,
-    Sending {
-        sender: mpsc::Sender<ExportTraceServiceRequest>,
-        // Node data should only be sent on the first message of a streaming
-        // request.
-        node: Option<Node>,
-        rsp: Option<
-            Pin<
-                Box<
-                    dyn Future<
-                            Output = Result<
-                                grpc::Response<Streaming<ExportTraceServiceResponse>>,
-                                grpc::Status,
-                            >,
-                        > + Send
-                        + 'static,
-                >,
-            >,
-        >,
-        metrics: Registry,
-    },
-}
-
 // ===== impl SpanExporter =====
 
 impl<T, S> SpanExporter<T, S>
 where
     T: GrpcService<BoxBody> + Clone,
-    S: Stream<Item = Span>,
+    T::Error: Into<Error>,
+    <T::ResponseBody as HttpBody>::Error: Into<Error> + Send + Sync,
+    T::ResponseBody: 'static,
+    S: Stream<Item = Span> + Unpin,
 {
-    const DEFAULT_MAX_BATCH_SIZE: usize = 100;
-
-    pub fn new(client: T, node: Node, spans: S, metrics: Registry) -> Self {
+    fn new(client: T, node: Node, spans: S, metrics: Registry) -> Self {
         Self {
             client,
             node,
             spans,
-            state: State::Idle,
-            max_batch_size: Self::DEFAULT_MAX_BATCH_SIZE,
             metrics,
         }
     }
 
-    fn do_send(
-        spans: Vec<Span>,
-        sender: &mut mpsc::Sender<ExportTraceServiceRequest>,
-        node: &mut Option<Node>,
-        metrics: &mut Registry,
+    async fn run(mut self) {
+        'reconnect: loop {
+            let (request_tx, request_rx) = mpsc::channel(1);
+            let mut svc = TraceServiceClient::new(self.client.clone());
+            let req = grpc::Request::new(request_rx);
+            trace!("Establishing new TraceService::export request");
+            self.metrics.start_stream();
+            let mut rsp = Box::pin(svc.export(req));
+            let mut drive_rsp = true;
+            loop {
+                tokio::select! {
+                    res = &mut rsp, if drive_rsp => match res {
+                        Ok(_) => drive_rsp = false,
```

Comment:
nit. this is probably just me but I find it easier to read this if the assigment is its own statement

```suggestion
                        Ok(_) => {
                            drive_rsp = false;
                        }
```